core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -749,25 +759,78 @@ private void updateStageExecutions()
 
         private boolean isReadyForExecution(SubPlan subPlan)
         {
-            for (SubPlan child : subPlan.getChildren()) {
-                StageExecution childExecution = stageExecutions.get(getStageId(child.getFragment().getId()));
-                if (childExecution == null) {
+            boolean speculative = false;
+            for (SubPlan source : subPlan.getChildren()) {
+                StageExecution sourceStageExecution = stageExecutions.get(getStageId(source.getFragment().getId()));
+                if (sourceStageExecution == null) {
+                    // source stage did not yet start
                     return false;
                 }
-                // TODO enable speculative execution
-                if (childExecution.getState() != StageState.FINISHED) {
+
+                if (sourceStageExecution.getState() != StageState.FINISHED) {
+                    if (exchangeManager.supportsConcurrentReadAndWrite()) {
+                        // speculative execution not supported by Exchange implementation
+                        return false;
+                    }
+                    speculative = true;
+                }
+                else {
+                    // source stage finished; no more checks needed
+                    continue;
+                }
+
+                if (!canOutputDataEarly(source)) {
+                    // no point in starting stage if source stage needs to complete before we can get any input data to make progress
+                    return false;
+                }
```

Comment:
@arhimondr I discussed this part with @linzebing and it feels that we should probably drop this check. At least in current shape.
Here we are filtering out final aggregations. But it is not true that final aggregation can not output data before stage complete. Individual task cannot, but when task completes we already have part of the output, even thought there can still be many tasks from aggregation stage in progress. 

What is your take on this one?

Reply:
Hmm, that's a good point. What do you think about change this check to something like `if (!canOutputDataEarly(source)) && !atLeastOneTaskIsFinished(source)`?

Reply:
`atLeastOneTaskIsFinished` is currently implicit - as we estimate output size based on some fraction of tasks complete. May change when I implement the "small input stage" flow. But currently we would not trigger downstream stage speculatively if at least a fraction of upstream tasks already completed execution.